File: lib/data/sources/local/daos/verse_dao.dart
import 'package:froom/froom.dart';

import '../../../entities/verse_entity.dart';

@dao
abstract class VerseDao {
  @Query(
    'SELECT * FROM verses WHERE bibleId = :bibleId AND chapterId = :chapterId ORDER BY verseNumber',
  )
  Future<List<VerseEntity>> getVersesByChapter(
    String bibleId,
    String chapterId,
  );

  @Query(
    'SELECT COUNT(*) FROM verses WHERE bibleId = :bibleId AND chapterId = :chapterId',
  )
  Future<int?> countVersesInChapter(String bibleId, String chapterId);

  @Insert(onConflict: OnConflictStrategy.replace)
  Future<void> insertVerses(List<VerseEntity> verses);

  @Query('DELETE FROM verses WHERE bibleId = :bibleId')
  Future<void> deleteVersesByBible(String bibleId);

  @Query('SELECT DISTINCT chapterId FROM verses WHERE bibleId = :bibleId')
  Future<List<String>> getDownloadedChapterIds(String bibleId);
}

File: core/database/src/main/java/com/biblelib/core/database/daos/VerseDao.kt
package com.biblelib.core.database.daos

import androidx.room.Dao
import androidx.room.Insert
import androidx.room.OnConflictStrategy
import androidx.room.Query
import com.biblelib.core.database.model.VerseEntity

@Dao
interface VerseDao {
    @Query("SELECT * FROM verses WHERE bibleAbbr = :abbr AND chapterId = :chapterId LIMIT 1")
    suspend fun getChapter(abbr: String, chapterId: String): VerseEntity?

    @Query("SELECT chapterId FROM verses WHERE bibleAbbr = :abbr")
    suspend fun getCachedChapterIds(abbr: String): List<String>

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun insert(verse: VerseEntity)

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    suspend fun insertAll(items: List<VerseEntity>)

    @Query("DELETE FROM verses WHERE bibleAbbr = :abbr")
    suspend fun deleteByBible(abbr: String)

    @Query("DELETE FROM verses")
    suspend fun deleteAll()

    @Query("""
        SELECT * FROM verses 
        WHERE bibleAbbr = :abbr AND contentJson LIKE '%' || :query || '%'
    """)
    suspend fun searchInBible(abbr: String, query: String): List<VerseEntity>
}